### PR TITLE
feat: Support directory paths in scans for Parquet, IPC and CSV

### PR DESCRIPTION
**Pull request by nameexhaustion**
This PR enables passing directory paths to `scan_(parquet|ipc|csv)`, which will be recursively traversed to load all files.

There is also some refactoring around hive partition logic to prepare for future behavior changes.

### CODE REVIEW

File: crates/polars-lazy/src/scan/file_list_reader.rs [hunk 1]
```diff
@@ -104,19 +205,9 @@ pub trait LazyFileListReader: Clone {
         true
     }
 
-    /// Path of the scanned file.
-    /// It can be potentially a glob pattern.
-    fn path(&self) -> &Path;
```
Comment (nameexhaustion) [inline crates/polars-lazy/src/scan/file_list_reader.rs:109]: Removed the `path` function from the trait - we just use a length-1 slice in `paths` for single paths

File: py-polars/tests/unit/io/test_scan.py [hunk 2]
```diff
@@ -191,9 +192,6 @@ def data_file(
 def test_scan(
     capfd: Any, monkeypatch: pytest.MonkeyPatch, data_file: _DataFile, force_async: bool
 ) -> None:
-    if data_file.path.suffix == ".csv" and force_async:
```
Comment (nameexhaustion) [inline py-polars/tests/unit/io/test_scan.py:194]: drive-by - enable these parametric tests for CSV now it we supports async

File: crates/polars-lazy/src/scan/file_list_reader.rs [hunk 3]
```diff
@@ -11,23 +12,124 @@ use crate::prelude::*;
...
+                    out_paths.push(path.map_err(to_compute_err)?);
+                }
+            } else {
+                out_paths.push(path.clone());
+            }
+        }
     }
+
+    // Todo:
+    // This maintains existing behavior - will remove very soon.
```
Comment (nameexhaustion) [inline crates/polars-lazy/src/scan/file_list_reader.rs:126]: will change this and add tests in a follow-up PR

File: crates/polars-plan/src/logical_plan/mod.rs [hunk 4]
```diff
@@ -80,6 +81,7 @@ pub enum DslPlan {
         paths: Arc<[PathBuf]>,
         // Option as this is mostly materialized on the IR phase.
         file_info: Option<FileInfo>,
+        hive_parts: Option<Vec<Arc<HivePartitions>>>,
```
Comment (nameexhaustion) [inline crates/polars-plan/src/logical_plan/mod.rs:84]: Moved this out from `file_info` - we store hive parts for every path in a `Vec` that we can resolve up-front - this replaces the current approach of using `update_hive_partitions`.

**Comment by ritchie46** [comment]
Can you do a rebase? :see_no_evil: The physical engine is moved to its own crate.

File: crates/polars-lazy/src/scan/csv.rs [hunk 5]
```diff
@@ -35,8 +34,7 @@ impl LazyCsvReader {
 
     pub fn new(path: impl AsRef<Path>) -> Self {
         LazyCsvReader {
-            path: path.as_ref().to_owned(),
```
Comment (ritchie46) [inline crates/polars-lazy/src/scan/csv.rs:38]: Nice! Having 2 of them was iffy at best. :smile:

**Comment by nameexhaustion** [comment]
rebased 👍

File: crates/polars-lazy/src/scan/file_list_reader.rs [hunk 6]
```diff
@@ -11,23 +12,124 @@ use crate::prelude::*;
 
 pub type PathIterator = Box<dyn Iterator<Item = PolarsResult<PathBuf>>>;
 
-// cloud_options is used only with async feature
-#[allow(unused_variables)]
-fn polars_glob(pattern: &str, cloud_options: Option<&CloudOptions>) -> PolarsResult<PathIterator> {
-    if is_cloud_url(pattern) {
+fn expand_paths(
```
Comment (ritchie46) [inline crates/polars-lazy/src/scan/file_list_reader.rs:15]: Maybe add a docstring explaining that this does globbing and/or recursive directory traversal.

File: crates/polars-lazy/src/scan/file_list_reader.rs [hunk 7]
```diff
@@ -145,23 +236,9 @@ pub trait LazyFileListReader: Clone {
...
-                Ok(None)
-            }
-        } else {
-            polars_ensure!(self.path().to_string_lossy() == "", InvalidOperation: "expected only a single path argument");
-            // Lint is incorrect as we need static lifetime.
-            #[allow(clippy::unnecessary_to_owned)]
-            Ok(Some(Box::new(paths.to_vec().into_iter().map(Ok))))
-        }
+    /// Returns a list of paths after resolving globs and directories, as well as
+    /// the hive start index.
```
Comment (ritchie46) [inline crates/polars-lazy/src/scan/file_list_reader.rs:240]: What does a `hive start index` mean?

**Review by ritchie46 — COMMENTED** [review]
Cool stuff. Left some questions.

File: crates/polars-lazy/src/scan/file_list_reader.rs (hunk 7 above)
Comment (nameexhaustion) [reply crates/polars-lazy/src/scan/file_list_reader.rs]: refers to where we start to parse hive partitions in the path - added comments here and above the function

File: crates/polars-lazy/src/scan/file_list_reader.rs (hunk 6 above)
Comment (nameexhaustion) [reply crates/polars-lazy/src/scan/file_list_reader.rs:18]: added
